### PR TITLE
Add Code of Conduct

### PR DESCRIPTION
**Pull request by fd0**
### What is the purpose of this change? What does it change?

 * It adds a code of conduct based on the one from the Django project.
 * It documents the current gorvernance model for this project

### Was the change discussed in an issue or in the forum before?

https://forum.restic.net/t/should-restic-adopt-a-code-of-conduct/445

### Checklist

- [X] I have read the [Contribution Guidelines](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#providing-patches)
- [x] There's a new file in `changelog/unreleased/` that describes the changes for our users (template [here](https://github.com/restic/restic/blob/master/changelog/changelog-entry.tmpl))
- [x] All commit messages are formatted in the same style as [the other commits in the repo](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#git-commits)
- [x] I'm done, this Pull Request is ready for review

### TODO

- [x] Decide what the enforcement rules are
- [x] Add email address

I'll squash all changes into one commit once all reviewers have signaled their consent. :)

### CODE REVIEW

File: GOVERNANCE.md [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+means. It is derived from the [i3 window manager project
+governance](https://raw.githubusercontent.com/i3/i3/next/.github/GOVERNANCE.md).
+
+## Roles
+
+* user: anyone who interacts with the restic project
+* core contributor: a handful of people who have contributed significantly to
+  the project by any means (issue triage, support, documentation, code, etc.).
+  Core contributors are recognizable via GitHub’s "Member" badge.
+* BDFL: a single individual who makes decisions when consensus cannot be
```
Comment (ifedorenko) [inline GOVERNANCE.md]: maybe expand `BDFL` abbreviation? I had to google it to make sure I knew what it meant precisely.

**Review by ifedorenko — APPROVED** [review]
+1

File: GOVERNANCE.md (hunk 1 above)
Comment (fd0) [reply GOVERNANCE.md]: oh, good point, thanks!

Comment (fd0) [reply GOVERNANCE.md]: Hm. It was already explained in the section above, but now it's in there twice. Doesn't hurt, I'd say.

**Comment by rawtaz** [comment]
I think this is very weird, that the CoC is put in the CONTRIBUTION document.

The contribution document is meant to explain how to contribute and what procedures to follow in doing so. It isn't the place for a CoC - the CoC should be its own document.

In general even if one argues that people who contribute should follow the CoC, this (the CONTRIBUTION file) is not the place to put that. We should not rub the CoC in the faces of people who might just want to submit code or other material that has very little to do with how we communicate and treat each other. IMO this is a sensitive subject and the guidelines for it it should be kept separate (and be just that, guidelines, not a "you must abide to this, otherwise you are not welcome to try contributing).

The bottom line; let's have the CoC be something that explains our expectations, not be a formal agreement that you must acknowledge that you adhere to just to be able to contribute something. *Even if people does not want to formally sign/acknowledge the CoC, this doesn't have to mean they will not follow it or that they would act in a way that would violate it.* For this reason, it does belong in a more "instructional" document like the CONTRIBUTION file, but is better placed in its own document.

**Comment by fd0** [comment]
Thanks for the comment, I've put the CoC into the file [recommended by GitHub](https://help.github.com/articles/adding-a-code-of-conduct-to-your-project/), `CODE_OF_CONDUCT.md`.

**Comment by rawtaz** [comment]
Not sure if it's effectively better in the end. The CONTRIBUTION file points to the CoC file and effectively tells people that they must follow this. But I guess that's what the project wants.

**Review by zcalusic — APPROVED** [review]

**Comment by fd0** [comment]
We'll try it this way, in a separate file with a hint in the `CONTRIBUTIONS.md`. We'll see how that goes. :)

File: GOVERNANCE.md [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+  when consensus cannot be reached. restic's current BDFL is [@fd0](https://github.com/fd0).
+
+## Decision making process
+
+In general, we try to reach consensus in discussions. In case consensus cannot
+be reached, the BDFL makes a decision.
+
+## Contribution process
+
+Please see [CONTRIBUTING](CONTRIBUTING.md).
```
Comment (fawick) [inline GOVERNANCE.md]: Maybe add a section how the CoC would be inforced and by whom? (source material e.g. here: https://opensource.guide/code-of-conduct/#enforcing-your-code-of-conduct)

Comment (fd0) [reply GOVERNANCE.md]: Hm. The Django enforcement manual feels a bit too much.

Comment (fd0) [reply GOVERNANCE.md]: I've added a section, please let me know what you think

Comment (fawick) [reply GOVERNANCE.md]: LGTM

**Review by fawick — APPROVED** [review]

**Comment by fd0** [comment]
Still waiting for @middelink, ping?

**Comment by rawtaz** [comment]
After pondering this for quite a while, I think that there are some real issues with this Code of Conduct (at least for those who care about it - for the majority that don't read it or that don't care about agreement-like texts or legalities, it'll be hard to explain the concerns, but I intend to at least give it a try).

For that reason I hereby ask that this PR is not merged until any and all concerns have been discussed and settled.

Unfortunately I do not have the time _right now_ to start elaborating on the issues, which is why I ask that the project please let this take its time. I'm not aware of any reason that this PR needs to be implemented quickly, so I'm hoping that this isn't a problem.

I'll try to start addressing the concerns in a few days time.

**Comment by cfcs** [comment]
@rawtaz I'd be interested to hear your concerns with this document :)

**Comment by alphapapa** [comment]
Strongly recommend against merging and imposing a code of conduct upon the project.

One of the most important reasons is that, by doing so, you are imposing a burden of enforcement upon the project leaders, and you are thereby implying that any conduct which is not punished is endorsed.  

It's similar to the difference between a common carrier and a publisher: a common carrier is not responsible for what content passes through it, and content passing through it does not imply endorsement; however, a publisher is responsible for all content it publishes.

We don't want the project leaders burdened by a duty to arbitrate silly squabbles.  The focus of the project should be the software.  

Recent history has shown that, in project after project, someone appears out of the woodwork to innocently suggest imposing a code of conduct.  It often turns out badly.  One of the reasons is that it can provide a platform for a future power grab.  

This is doubly so when the code of conduct does not explicitly limit itself to official project spaces.  This is, in fact, the case with the proposed code:

> This code of conduct applies to all spaces managed by the restic project. This includes IRC, the issue tracker, the forum, and any other forums created by the project team which the community uses for communication. In addition, **violations of this code outside these spaces may affect a person's ability to participate within them.**

This is a very dangerous precedent to set.  We have seen in recent years how codes of conduct exactly like this have led to the expulsion of valued project members because of comments they have made, e.g. on their Twitter accounts, completely irrelevant to the projects they participated in and led.  The highlighted phrase in the quoted passage is a ticking time bomb.  The idea that project leaders are responsible for policing participants' speech in other forums is a bizarre and terrible idea.

The best thing to do would be to abandon the idea of a code of conduct, policing, enforcement, etc.  Let those who want to participate do so.  Let those who don't, don't.  Let the project leader handle issues as he sees fit.  Let those who don't like his leadership fork the project.

At most, a code of conduct should be something like:

> This is a software project.  Any issues or differences of opinion, which are not about the software, between participants, are not relevant to the project.  The project does not discriminate against participants for any issues not directly related to the software.  Any personal differences of opinion unrelated to the project should be settled privately between participants.  Participants who are unwilling to work with other participants are encouraged to fork the project.

Anything more is setting the project up for yet another round of drama and headlines on Hacker News and Gizmodo, etc, a hostile fork, etc.  We don't need or want this kind of stuff in a project about backup software.  We don't want personal squabbles, political infighting, factions and tribalism, from irrelevant Twitter posts.  We just want reliable backup software that doesn't care about who makes it or uses it.

**Comment by tamalsaha** [comment]
@fd0 , if you don't mind me asking, why the sudden (just my observation) impetus for adding coc?.

I think languages like "In addition, violations of this code outside these spaces may affect a person's ability to participate within them." are not conducive to free thinking and open discussion (where people might not always agree).

If coc must be added, I would encourage taking a look at CNCF's coc doc: https://github.com/cncf/foundation/blob/master/code-of-conduct.md

This document is much more focused on the project space.

**Comment by MarkMielke** [comment]
I feel like this is a losing battle, so I haven't commented on the many other projects I have seen this happen to. But, I have to add the comment that I too don't get it. For techno-geeks like me, Github and technical projects like Restic have felt like a refuge from the ridiculous political correctness efforts. Where we can meet on logic and code and shared passion, and nothing else. I don't care if you are male or female, black or white, rich or poor. I only care what you contribute, and the words that you write. But now I'm being asked to both care, and be subject to, some code of conduct which seems written with explicit intent to bring about, and enforce a particular way of behaving that I was not consulted on?

I haven't yet contributed to Restic, but I have contributed to others. I've been watching you for months now, and learning about Restic, and when we do start using it, I do plan on contributing. I think this community is healthy. It feels like explicitly adding a code of conduct is in response to something unhealthy, which feels like there is something being judged. In other open source projects that have gone down this path, there has been at least one high profile person quitting over this, and I feel like I understand his motivation. I don't want to be judged by a code of conduct. I want you to judge my words by their merit alone, and not by whether it reflects some agenda that is external from this project. You don't need a code of conduct to kick out bad people, and you don't need a code of conduct to encourage good behaviour. Bad people don't care about your code of conduct in the first place, and good people are already doing it.

I also have greater respect for people who treat bad behaviour with tolerance, vs people who point to a rule book and eject the offender. There is usually a reason behind the bad behaviour, and getting to the bottom of it can be important.

The words themselves are not really objectionable. But, the motivation behind it is objectionable to me. I particularly object to how anybody who dares challenge these is branded as clearly somebody defending bad behaviour. This exact reaction is what I find objectionable.

Adding political and social aspects to the project makes me less comfortable here. Will I leave? Probably not. Restic is valuable software that I would like to hopefully contribute to in the future. If it came down to it, I would swallow my objections. But, here is the opportunity for me to register my objections.

This document concerns me, and this path concerns me.

**Comment by rawtaz** [comment]
@cfcs I should have gotten back on this earlier, I feel a bit bad about that. I've had a lot to attend lately. I'm going to do it ASAP. Funny though, other people just the last day wrote better texts on this topic than I ever could!

What @MarkMielke wrote is one of the best and most well-written comments I have ever read, and it encapsulates the absolute majority of my view on this as well. Especially:

> You don't need a code of conduct to kick out bad people, and you don't need a code of conduct to encourage good behaviour. Bad people don't care about your code of conduct in the first place, and good people are already doing it.

Let the individuals and the community regulate itself, that's not a concern the project or its maintainers should have. But also, the other things he wrote. Bulls eye.

FWIW I already know people that with 99% certainty will stop contributing to the project if the current CoC and Governance is passed. I think that's too bad. The extremely few times there's some kind of conflict between individuals, to the level that it's even worth mentioning, is a way smaller problem than the cost of losing contributions is or would be.

**Comment by cfcs** [comment]
@MarkMielke I'll respond to you instead of @rawtaz since he expressed that he shares your take on it.

> "But now I'm being asked to both care, and be subject to, some code of conduct which seems written with explicit intent to bring about, and enforce a particular way of behaving that I was not consulted on?"
1. As the author of restic, Alex could just have pushed this to master. The point of this PR is exactly to consult the community and ask people's opinions.

> "Bad people don't care about your code of conduct in the first place, and good people are already doing it."
2. I thought the point of trying to write a code of conduct down here was to address what "bad behaviour" constituted, as a guide to good people so they have a better idea of how the interaction culture around the project is (as opposed to defaulting to the mailing-list culture of the 90's where it's every man for himself in a constant dick measuring contest).
I've been on plenty of those mailing-lists, I've sat in the IRC channels, I've argued with the megalomaniac misunderstood geniuses. Frankly this place is much nicer, and I would like it if it stayed that way.
A policy that defines some things as examples of bad behaviour makes it easier as a community to deal with those issues when they arise instead of having to have a massive debate about who's right when someone crosses the line, so the focus can be on how to make things right again.

> "I also have greater respect for people who treat bad behaviour with tolerance, vs people who point to a rule book and eject the offender. There is usually a reason behind the bad behaviour, and getting to the bottom of it can be important."
3. I couldn't agree more with this; social rejection rarely helps anyone. I don't see how this is orthogonal to anything mentioned in this PR though; I don't think any of the text is set in stone, so feel free to propose augmentations. While I honestly struggle to make sense of the rest you wrote, I think that you bring up an important point in this paragraph.

> Adding political and social aspects to the project makes me less comfortable here.
I just re-read the document again I don't see any mention of politics, could you elaborate?
Regarding "adding social aspects to the project" making you less comfortable, I'm not sure how you envision building a community desolate of social aspects?

I'm beginning to feel that maybe calling this a "code of conduct" has triggered some kind of allergic reaction among a certain segment of software developers. Maybe we should change the headline to "contributor guidelines" or something less likely to mobilize the reddit hordes?

**Comment by kurin** [comment]
I think that it's better for communities over a certain size to have a CoC than not to have one. In the first place it demonstrates a community's commitment to openness, and in the second place it gives people who don't know each other well and who might not otherwise interact, and who may be from different cultures with different norms, a touchstone for what kind of interactions to aspire to.

I also don't buy that it would tie anybody's hands when it comes to community moderation. It doesn't say so-and-so must be immediately expelled if a thread gets hot.

But your example of Linus, well... As someone who _does_ contribute to restic, if fd0 started yelling at me (or anyone else) the way Linus yells, I'd be gone yesterday, and frankly I appreciate a document that spells out explicitly that nobody should do that.

**Comment by MarkMielke** [comment]
@kurin:

For: "I also don't buy that it would tie anybody's hands when it comes to community moderation. It doesn't say so-and-so must be immediately expelled if a thread gets hot."

How about being explicit about this, then? Remember that ambiguity means that the strictest interpretation may be expected. What will the enforcement model be? As somebody mentioned prior, it was curiously specific about:

"In addition, violations of this code outside these spaces may affect a person's ability to participate within them."

You say it doesn't immediately expel people for direct moderation, and yet it explicitly states that it may expel people even for actions outside the the listed spaces entirely?

For the Linus example... I wonder if you are aware of exactly how bad the Intel Meltdown/Spectre issue is, what impacts it has had on the world, and how Intel has been mishandling it from the start, and continues to mishandle it in the Linux development discussions. To me, being open, is for everybody to say what everybody is thinking, which is truly WTF. Not being able to say WTF, is the opposite of open. It is closed. It is saying that WTF is an invalid reaction to a situation that only deserves WTF.

**Comment by kurin** [comment]
@MarkMielke:

If you want to propose changes to the language I don't think anyone would have a problem with that.  The code as written takes into account behavior that happens "outside these spaces", yes, but my point is that this _sets expectations_, it doesn't mandate a response or what that response should be.

> I wonder if you are aware of exactly how bad the Intel Meltdown/Spectre issue is, [...] To me, being open, is for everybody to say what everybody is thinking, which is truly WTF.

I think it doesn't matter how bad Spectre is, or how badly Intel bungled the response.  The way Linus treats people, sometimes, is unacceptable.  I wouldn't accept it from a colleague, a boss, or an anonymous internet collaborator.  I think it's toxic, I'm _positive_ it's driven more talent away from the Linux project than Linus' considerable talent justifies, and I am absolutely behind efforts, like this, that spell that out explicitly.

Exhortations to "be civil" or "be professional" don't prevent anyone from calling out bad code, they just (ought to) prevent people from being _mean_ about it.

File: CODE_OF_CONDUCT.md [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+connecting people.
+
+Diversity is one of our huge strengths, but it can also lead to communication
+issues and unhappiness. To that end, we have a few ground rules that we ask
+people to adhere to. This code applies equally to founders, mentors and those
+seeking help and guidance.
+
+This isn't an exhaustive list of things that you can't do. Rather, take it in
+the spirit in which it's intended - a guide to make it easier to enrich all of
+us and the technical communities in which we participate.
```
Comment (MarkMielke) [inline CODE_OF_CONDUCT.md]: Does the above paragraph add substantial value to the document? I feel like it distracts from the rest and it isn't clear whether it suggests an increase or decrease in intent to enforce. "Isn't exhaustive of things that you can't do" implies that that it might encompass more than just written. "Guide to make it easier" suggests that it might be overly literal and simplistic, and is actually intended for reasonable interpretation.

File: CODE_OF_CONDUCT.md [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+
+This code of conduct applies to all spaces managed by the restic project. This
+includes IRC, the issue tracker, the forum, and any other forums created by the
+project team which the community uses for communication. In addition,
+violations of this code outside these spaces may affect a person's ability to
+participate within them.
```
Comment (MarkMielke) [inline CODE_OF_CONDUCT.md]: As a general principle, I don't like the idea that the code of conduct for one project would enforce behaviour in other forums where the code of conduct might be different. It's somewhat like one country defining rules of conduct in other countries. This is confusing. I expect this is to cover extreme cases such as when somebody has done something so troublesome as to make it a concern that their conduct in another forum might in any way reflect poorly on this project.

I think there is a wide range of possible concerns here, and the starting point should be to leave it out entirely. However, if you want to cover this point - perhaps being more explicit would help avoid this being a concern: "Although it is not the responsibility of the restic project to define and enforce code of conduct outside of the restic project, there may be situations that require the restic project to distance itself from the actions of a contributor. In this unfortunate situation, a contributor may be asked to leave the community so as to preserve the good reputation of the project."

File: CODE_OF_CONDUCT.md [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+
+If you believe someone is violating the code of conduct, we ask that you report
+it by emailing conduct AT restic.net.
+
+ * Be friendly and patient.
```
Comment (MarkMielke) [inline CODE_OF_CONDUCT.md]: I believe there is a section header missing from above this list, that would set the context of what follows. Perhaps "Guiding Principles:" would be a good lead-in so as to suggest these are intended to be read as a well-intended list of principles, but without implying that the list is exhaustive, or that the list should be taken too literally? It really serves as a model for what good behaviour the community should target.

**Review by MarkMielke — CHANGES_REQUESTED** [review]

**Review by MarkMielke — COMMENTED** [review]
@kurin:

I took your advice and provided suggested changes.

My general concern is really with the words being taken too literally, and with too much potential for somebody to manipulate the system. I much prefer the model where the BDFL and two core contributors decides what the reasonable thing to do is, and do not have their hands tied by text that is taken too literally. So, my suggestions are to slightly soften the text, without getting rid of the valuable "good behaviour" definitions.

I still think there are situations that are so bad as to allow for honest human reactions of dismay or disgust, that could be offensive. The Intel Meltdown/Spectre is an example of this in that it has caused massive world-wide damage that has still not been fully felt, stretching even to Restic, as any system patched to mitigate against these is likely running 2% - 20% slower today, depending upon the workload and any system not yet patched may be open to attack and exposure of all private information on the system by JavaScript running in a web browser. This issue is truly worthy of being professionally disgusted, in whatever form that take, just as you would be professionally disgusted if an architect incorrectly designed a building that fell down and killed thousands of people. These are extreme cases, and we shouldn't normally experience such extreme cases in real life. But, in the unlikely event that core contributors to Restic starting submitting and approving untested patches and released them, and this caused user systems to broadly lose backup data... and if it was my data... I would start to swear. And anybody who challenged me on my right to do this would get an earful. And if I was kicked out of the project for it, I would fork it and part ways with the broken community. That all said... I have no reason to believe this would ever happen. It's just a theoretical to explain under what situation all of this might occur.

File: CODE_OF_CONDUCT.md (hunk 3 above)
Comment (alphapapa) [reply CODE_OF_CONDUCT.md]: You're right, and it's even worse than that.  The first sentence is an open door to expelling anyone from the project for nearly any reason, leaving anyone's behavior open to interpretation according to whoever's doing the interpreting.  Proscribed behaviors don't have to be on the list to be forbidden.  Whoever's in power decides what's allowed, which also allows them to decide unfairly among those they like and dislike.

The second sentence is a rhetorical trap: how could anyone object to that?  Anyone who objects to it must be against the spirit of enriching the community.  Therefore their objections are invalid (and they should probably be expelled after the code is imposed, to protect and enrich the community, of course).

These illustrate why the wholesale adoption of pre-packaged codes of conduct is a very bad idea.  It's like a water fountain that's been set up with a sign that says, "Help yourself!  Be enriched!"  Seems like a great, generous idea--yet the water is poisoned, and those (i.e. projects) who drink of it become infected.

And how, then, would the infection be eradicated?  Who could ever *erase* a code of conduct?  Whoever would want to do that must be a *toxic* person who doesn't care about the health of the community.  Begone, evildoer.

File: CODE_OF_CONDUCT.md (hunk 4 above)
Comment (alphapapa) [reply CODE_OF_CONDUCT.md]: > As a general principle, I don't like the idea that the code of conduct for one project would enforce behaviour in other forums where the code of conduct might be different. It's somewhat like one country defining rules of conduct in other countries. 

You mean like the GDPR?  ha...ha...

> I think there is a wide range of possible concerns here, and the starting point should be to leave it out entirely. However, if you want to cover this point - perhaps being more explicit would help avoid this being a concern...

I appreciate your trying to be reasonable, but it's still a very bad idea, no matter how nicely or carefully you word it.  It's just a ticking time bomb.

This is backup software.  We just want good software.  That should be our only goal here.  Politics and societal issues ought to be explicitly barred from being relevant to the project and participation in it.

**Comment by fd0** [comment]
Thank you all for taking the time to write extensive comments. I'd like to split out the governance part and push this to master, it just documents the status quo.

We'll see what we come up with in terms of the conduct. I like the term "community guidelines"
